### DISCUSSION TITLE
Trouble launching a .py file

### DISCUSSION THREAD

**Discussion by ISIR-GuillaumeS** [sofa-framework/sofa#4021] [Build / Config / Environment] [answered]
Hi !

I try to launch an example (the cablegripper one) which is a python file. The error was telling that no python interpreter was launched (even if I do have the plugin). So I went to Edit ->Plugin Manager and add [libSofaPython3.so](http://libsofapython3.so/)
So now I still have an error which is :

Unable to completely load the scene from file '/home/benamar/Documents/Travail/SimuSofa/details/[step1.py](http://step1.py/)'.  
Python exception:   
  ModuleNotFoundError: No module named 'stlib3'

At:
  /home/benamar/Documents/Travail/SimuSofa/details/[step1.py](http://step1.py/)(2): <module>
  <frozen importlib._bootstrap>(219): _call_with_frames_removed
  <frozen importlib._bootstrap_external>(848): exec_module
  <string>(6): <module>

Do you have an idea ?

**Comment by bakpaul** [comment, marked as answer]
Hi, stlib3 is a Defrost team plugin that can be found here : https://github.com/SofaDefrost/STLIB. Have you installed it ?

**Reply by ISIR-GuillaumeS** [reply]
Hi ! Thanks for your answer !
Yes I already installed it.

**Reply by bakpaul** [reply]
Ok, and did you make an In-Tree compilation of the module or a out-of-tree one ?

**Reply by ISIR-GuillaumeS** [reply]
I did an In-Tree compilation. Do you thin that is the reason ?
I did not get how the out-of-tree compilation works

**Reply by bakpaul** [reply]
Actually I'm am not a python user with Sofa, I usually use xml files. But I just wanted to know how you built everything. I know that there is a similar issue on Defros's github that hasn't been answered yet. Maybe @alxbilger could be of help here. 

Until then, I may suggest for a workaround to modify your PYTHONPATH by adding the stilb3 library path to it. It may work.

**Reply by ISIR-GuillaumeS** [reply]
Alright thank you ! 
Could you explain me exactly how can I do it ? I've never done this before

**Reply by bakpaul** [reply]
I'm an ubuntu user so for me it is quite straightforward. In windows I found this on the web : ` set PYTHONPATH=%PYTHONPATH%;C:\path\to\your\module`

**Reply by ISIR-GuillaumeS** [reply]
I also am a ubuntu user but in fact I don't know so much of it x)

**Reply by bakpaul** [reply]
Ah even easier ! Add this to you `~/.bashrc` : 
`export PYTHONPATH=$PYTHONPATH:/path/to/plugins`
After that in order for the changes to be applied in any terminal that is open, you'll need to do a `source ~/.bashrc` or open a new one.

**Reply by ISIR-GuillaumeS** [reply]
Alright !! Does the "/path/to/plugin" is the path of the repertory to the external plugins that I add ?

**Reply by bakpaul** [reply]
Yes it is the absolute path of the folder containing the .so you want to load in your python scenes

**Reply by ISIR-GuillaumeS** [reply]
Isn't the same thing to got in Sofa to Edit->Plugin Manager and to add the lib needed ?

**Reply by bakpaul** [reply]
No. This will tell your python interpreter where to look for the modules when you do an import. It defines the environment. So it is more global and python related. When you add a lib in the plugin manager, it load its content to the current process. I am not a specialist in this domain, but I am not sure that this would affect the python environment the way we think.

**Reply by ISIR-GuillaumeS** [reply]
In fact I think I did something wrong during the configuration so I tried something else and here is what I got when I run ninja : 
ninja: error: manifest 'build.ninja' still dirty after 100 tries

Do you have an idea ?

**Reply by bakpaul** [reply]
Nope, no idea sorry ! If I were you, I would try recompiling from scratch...

**Reply by ISIR-GuillaumeS** [reply]
If I may come back to adding the path to bashrc.
I tried but I have some difficulties. I opened it and add at the end : 
" export PYTHONPATH=$PYTHONPATH:"/home/benamar/ext_plugin_repo/SofaPython3" "
Where /home/benamar/ext_plugin_repo/SofaPython3 is the path to the SofaPython3 Plugin v23.06
Is it right ? Because it doen't change anything :/

**Reply by bakpaul** [reply]
I would remove the quotes in the path and you should put the path to any lib you want to import in python. So if this folder contain the STLIB so then this is what I ment, ohterwise I suggest to add the path to this particular folder

**Reply by ISIR-GuillaumeS** [reply]
Oh Ok so that folder is here : 
![image](https://github.com/sofa-framework/sofa/assets/135211551/651ea791-15bb-4671-8a39-50c89f4a2b93)
Therefore I had this in my bashrc file :
export PYTHONPATH=$PYTHONPATH:/home/benamar/ext_plugin_repo/STLIB/python3/src
But that still don't work

**Reply by bakpaul** [reply]
No the folder should contain the .so files at its root. For instance my PYTHONPATH looks like that : 
`export PYTHONPATH=$PYTHONPATH:/home/paul/dev/build/flow_imgui/install/plugins`
And this folder contains : 
```
-rwxrwxr-x 1 paul paul   6950736 Mar 13 17:56 libros2_pybinding_general.so*
-rwxrwxr-x 1 paul paul   8653024 Mar 13 17:56 libros2_pybinding_sofa.so*
-rwxrwxr-x 1 paul paul   2219136 Mar 13 17:56 libros2_pylib_data.so*
-rwxrwxr-x 1 paul paul   1741776 Mar 13 17:56 libros2_pylib.so*
-rwxrwxr-x 1 paul paul   2462072 Mar 13 17:56 libros2_pylib_task.so*
-rwxrwxr-x 1 paul paul    768040 Mar 13 17:56 libros2_pylib_widget.so*
```
which are the modules I want to import in python 
So this folder should be a folder in the build directory where the generated library file (.so file) of the STLIB is located

**Reply by ISIR-GuillaumeS** [reply]
The problem is that I don't know what module I need. i just want to launch a python file so what am I searching for ?

**Reply by bakpaul** [reply]
I've just compiled it and found it on my computed under `/home/paul/dev/build/sofa/lib/libSTLIB.so` where `/home/paul/dev/build/sofa/` is the build directory of sofa. So you should try the same

**Reply by ISIR-GuillaumeS** [reply]
Ok I have tried some and it finally works !! I don't really know why and how I did it but still x)
Thank you so much for your help and time !!

**Reply by bakpaul** [reply]
No problem ! When dealing with execution or compilation failure like that, knowing which path is used, when, searching what and by whom is very important ! 
Enjoy !
